zeth-contracts/contracts/Groth16Mixer.sol

```diff
@@ -43,31 +83,274 @@ contract Groth16Mixer is BaseMixer {
+        // (len(givenInputs) needs to equal expectedInputSize (not less))
+        require(
+            input.length + 1 == verifyKey.ABC.length,
+            "Input length differs from expected");
+
+        // A memory scratch pad
+        uint[24] memory pad;
+
+        // 1. Compute the linear combination
+        //   vk_x = \sum_{i=0}^{l} a_i * vk.ABC[i], vk_x in G1.
+        //
+        // ORIGINAL CODE:
+        //   Pairing.G1Point memory vk_x = vk.ABC[0]; // a_0 = 1
+        //   for (uint i = 0; i < input.length; i++) {
+        //       vk_x = Pairing.add(vk_x, Pairing.mul(vk.ABC[i + 1], input[i]));
+        //   }
+        //
+        // The linear combination loop was the biggest cost center of the mixer
+        // contract.  The following assembly block removes a lot of unnecessary
+        // memory usage and data copying, but relies on the structure of storage
+        // data.
+        //
+        // `pad` is layed out as follows, (so that calls to precompiled
+        // contracts can be done with minimal data copying)
+        //
+        //  OFFSET  USAGE
+        //   0x20    accum_y
+        //   0x00    accum_x
+
+        // In each iteration, copy scalar multiplicaation data to 0x40+
+        //
+        //  OFFSET  USAGE
+        //   0x80    input_i   --
+        //   0x60    abc_y      | compute abc[i+1] * input[i] in-place
+        //   0x40    abc_x     --
+        //   0x20    accum_y
+        //   0x00    accum_x
+        //
+        //  ready to call bn256ScalarMul(in: 0x40, out: 0x40).  This results in:
+        //
+        //  OFFSET  USAGE
+        //   0x80
+        //   0x60    input_i * abc_y  --
+        //   0x40    input_i * abc_x   |  accum = accum + input[i] * abc[i+1]
+        //   0x20    accum_y           |
+        //   0x00    accum_x          --
+        //
+        //  ready to call bn256Add(in: 0x00, out: 0x00) to update accum_x,
+        //  accum_y in place.
+
+        bool success = true;
+        assembly {
+
+            let g := sub(gas, 2000)
+
+            // Compute slot of ABC[0], using pad[0] as a temporary
+            mstore(pad, add(verifyKey_slot, 10))
+            let abc_slot := keccak256(pad, 32)
+
+            // Compute input array bounds (layout: <len>,elem_0,elem_1...)
+            let input_i := add(input, 0x20)
+            let input_end := add(input_i, mul(0x20, mload(input)))
+
+            // Initialize pad[0] with abc[0]
+            mstore(pad, sload(abc_slot))
+            mstore(add(pad, 0x20), sload(add(abc_slot, 1)))
+            abc_slot := add(abc_slot, 2)
+
+            // Location within pad to do scalar mul operation
+            let mul_in := add(pad, 0x40)
+
+            // Iterate over all inputs / ABC values
+            for
+            { }
+            lt(input_i, input_end)
+            {
+                abc_slot := add(abc_slot, 2)
+                input_i := add(input_i, 0x20)
+            }
+            {
```

Comment:
Let's change the shape of this block to something closer to what we are used to write in higher level languages:

We can either go for something like:
```solidity
// Iterate over all inputs / ABC values
for
     { }
     lt(input_i, input_end)
     {
          abc_slot := add(abc_slot, 2)
          input_i := add(input_i, 0x20)
     }
{
   FOR BLOCK
}
```

Or:
```solidity
for { } lt(input_i, input_end) { abc_slot := add(abc_slot, 2); input_i := add(input_i, 0x20)}
{
   FOR BLOCK
}
```

Reply:
Chnaged to this layout.